compiler/rustc_mir_dataflow/src/impls/liveness.rs

```diff
@@ -179,3 +189,139 @@ impl DefUse {
         }
     }
 }
+
+/// Like `MaybeLiveLocals`, but does not mark locals as live if they are used in a dead assignment.
+///
+/// This is basically written for dead store elimination and nothing else.
+///
+/// All of the caveats of `MaybeLiveLocals` apply.
+pub struct MaybeTransitiveLiveLocals<'a, 'tcx> {
+    always_live: &'a BitSet<Local>,
+    local_decls: &'a LocalDecls<'tcx>,
+    tcx: TyCtxt<'tcx>,
+}
+
+impl<'a, 'tcx> MaybeTransitiveLiveLocals<'a, 'tcx> {
+    /// The `always_alive` set is the set of locals to which all stores should unconditionally be
+    /// considered live.
+    ///
+    /// This should include at least all locals that are ever borrowed.
+    pub fn new(
+        always_live: &'a BitSet<Local>,
+        local_decls: &'a LocalDecls<'tcx>,
+        tcx: TyCtxt<'tcx>,
+    ) -> Self {
+        MaybeTransitiveLiveLocals { always_live, local_decls, tcx }
+    }
+}
+
+impl<'a, 'tcx> AnalysisDomain<'tcx> for MaybeTransitiveLiveLocals<'a, 'tcx> {
+    type Domain = ChunkedBitSet<Local>;
+    type Direction = Backward;
+
+    const NAME: &'static str = "transitive liveness";
+
+    fn bottom_value(&self, body: &mir::Body<'tcx>) -> Self::Domain {
+        // bottom = not live
+        ChunkedBitSet::new_empty(body.local_decls.len())
+    }
+
+    fn initialize_start_block(&self, _: &mir::Body<'tcx>, _: &mut Self::Domain) {
+        // No variables are live until we observe a use
+    }
+}
+
+struct TransferWrapper<'a>(&'a mut ChunkedBitSet<Local>);
+
+impl<'a> GenKill<Local> for TransferWrapper<'a> {
+    fn gen(&mut self, l: Local) {
+        self.0.insert(l);
+    }
+
+    fn kill(&mut self, l: Local) {
+        self.0.remove(l);
+    }
+}
+
+impl<'a, 'tcx> Analysis<'tcx> for MaybeTransitiveLiveLocals<'a, 'tcx> {
+    fn apply_statement_effect(
+        &self,
+        trans: &mut Self::Domain,
+        statement: &mir::Statement<'tcx>,
+        location: Location,
+    ) {
+        // Compute the place that we are storing to, if any
+        let destination = match &statement.kind {
+            StatementKind::Assign(assign) => {
+                (|| {
+                    if let Rvalue::Cast(CastKind::Misc, src_op, dest_ty) = &assign.1 {
+                        // Pointer to int casts may be side-effects. While the model is undecided,
+                        // we should be conservative.
```

Comment:
Maybe explicitly talk about "expose"ing the pointer here, so that people have something to search for if they want to learn more about their side-effects. Or link to my [blog post](https://www.ralfj.de/blog/2022/04/11/provenance-exposed.html). ;)

Reply:
Are there any backends that use that semantics? Miri? If it isn't really used, I wouldn't go out of the way to preserve those cast. Supporting this semantics requires more comprehensive changes anyway.

Reply:
Miri uses (or soon will use) those semantics.

The other backends do not have a semantics for this operation (or the semantics they have is not internally consistent / not consistent with the optimizations they want to do).

So, I would definitely prefer treating ptr2int casts as side-effecting in MIR. (int2ptr casts should also be treated as impure, in the sense that repeating the operation multiple times can produce different results.)

Reply:
Ok. `SimplifyLocals` is another transformation that will remove pointer to integer casts when destination local is never used anywhere else.

Reply:
Ah, IMO that should be changed then.

Maybe they should become a different `CastKind` to make them easier to distinguish.

Reply:
I mean, if we really are going to make them side effects then they should become a terminator imo

Reply:
I don't follow. Assignment is a side-effecting statement, since it changes the contents of memory.

Reply:
Yeah, "side-effecting" is too overloaded now. What I mean is "statements whose results are not captured in their 'output'"

Reply:
I would say that is definitely the case in `*foo = value`. Or in `Retag(place)`. The latter doesn't even have an output.

I don't see how these casts are any worse than that. The goal here is to make the life easier for analyses, right? We already cannot freely reorder statements inside a basic block. So I don't see what we gain by making casts into terminators.

Reply:
Hmm, it still feels like a distinction, but I'm happy to delay this until I have more concrete examples of what I mean

Reply:
With regards to making them a `CastKind`: I agree that we should do this, but I'm not going to do it right now. I need to understand why `CastKind` is the way it is right now anyway (I'm about to go ask on Zulip) and then possibly refactor it more generally

Reply:
I opened #97421 to track the issue of pointer to integer casts removed by SimplifyLocals.